### PR TITLE
HOMME: Kokkos implementation of theta-l target

### PR DESCRIPTION
**Pull request by bartgol**
This PR brings in the kokkos implementation of the theta-l target in HOMME. It is the implementation that we used for the 2020 Supercomputing/Gordon-Bell submission (paper accepted, not GB), with very few modifications following the merge of master into this branch.

After the merge, I manually verified that our unit tests (CXX vs F90 bfb comparison) pass on SKX and V100 architectures (KNL still to do). I am currently checking that also end-to-end tests are bfb (cxx vs f90), when run in debug mode. I also manually verified that the performance on V100 of the current branch (after master merge) has not changed compared to the version we used for the SC/GB submission.

Note: this PR also upgrades the version of kokkos in our external submodule. As a consequence, some machine files have to be updated (I haven't gotten to that yet).

[nonbfb] for homme suites (nonbfb suite -- sl executables; bfb suite - f theta executables, 1 sl test, new baselines needed)
[bfb] for EAM

### CODE REVIEW

File: components/cam/bld/configure [hunk 1]
```diff
@@ -2658,6 +2658,7 @@ sub write_filepath
       }
       elsif($opts{'dyn_target'} eq 'theta-l') {
         print $fh "$camsrcdir/homme/src/theta-l\n";
+        print $fh "$camsrcdir/homme/src/theta-l/share\n";
```
Comment (oksanaguba) [inline components/cam/bld/configure:2661]: I'd suggest keeping this PR homme-only, no cam changes (which should be the next PR).

Comment (bartgol) [reply components/cam/bld/configure:2661]: well, without that change, cam won't find some files. In order for theta-l_kokkos to reuse some f90 files from theta-l, I did what was done for preqx, and created the share folder (I think we discussed it on slack, and that was the agreed upon strategy). So some files have been moved from theta-l to theta-l/share.

File: components/homme/test_execs/preqx-nlev72-kokkos/CMakeLists.txt [hunk 2]
```diff
@@ -14,11 +11,17 @@ createTestExec(preqx-nlev72-kokkos preqx_kokkos 4 4 72 FALSE FALSE 35)
 SET (HOMME_TEST_VCOORD_INT_FILE acme-72i.ascii)
 SET (HOMME_TEST_VCOORD_MID_FILE acme-72m.ascii)
 SET (HOMME_TEST_NE 2)
+SET (HOMME_TEST_TIME_STEP 300)
 SET (HOMME_TEST_NDAYS 1)
 SET (HOMME_TEST_QSIZE 4)
 SET (HOMME_TEST_RSPLIT 3)
 SET (HOMME_TEST_MOISTURE notdry)
 SET (HOMME_TEST_TIME_STEP 600)
+SET (HOMME_TEST_HVSCALING 0)
```
Comment (oksanaguba) [inline components/homme/test_execs/preqx-nlev72-kokkos/CMakeLists.txt:20]: why do you need these changed for preqx xx ?

File: components/cam/bld/configure (hunk 1 above)
Comment (oksanaguba) [reply components/cam/bld/configure:2661]: Is it needed to keep CAM target theta-l functional? if not, let's keep it out of this PR. btw I will run cam tests anyway before this is merged.

File: components/homme/test_execs/preqx-nlev72-kokkos/CMakeLists.txt (hunk 2 above)
Comment (bartgol) [reply components/homme/test_execs/preqx-nlev72-kokkos/CMakeLists.txt:20]: I think I added it cause the `preqx.nl` template namelist used to generate the tests has that setting too, so I didn't want to have entries in the namelist that look like `blah =` (without a rhs). Is that param not used by preqx?

**Comment by bartgol** [comment]
@jgfouca I added you as a reviewer, since this PR includes a micro-mod in e3sm's cmake logic, albeit limited to the handling of kokkos (see last commit).

**Review by jgfouca — APPROVED** [review]

**Comment by oksanaguba** [comment]
@bartgol I want to merge this pr this week. right now there are some conflicts, but when I tried to merge to the next some other subset of conflicts appeared (conflicts in imex mod were gone). so, before merge I might submit a commit with conflict resolves, so that I don't resolve conflicts twice.

**Comment by bartgol** [comment]
@oksanaguba Whatever is easier for you, works for me. Let me know if/how I can help with that.

**Comment by oksanaguba** [comment]
So far status: After merging to next from yesterday and running against the same next on anvil+intel:
homme suite -- some unit tests fail, probably due to avx, fix is on next
hommebfb suite -- all theta F tests tail, prob due to T init
atm_integration -- all pass

**Comment by rljacob** [comment]
Can be merged.  No impact on WC.

**Comment by oksanaguba** [comment]
Verifying the current state: I merged this branch to next from yesterday and compared with the same next.

For HOMME nonbfb suite diffs are in

/home/ac.onguba/as/HOMME_P24.f19_g16_rx1.A.anvil_intel.C.20201008_234625_gl6kf4

```
21/23 Test #21: thetah-sl-test11conv-r1t2-cdr20 .......***Failed    2.22 sec
Submitting 1 jobs
Running test thetah-sl-test11conv-r1t2-cdr20-run ... /lcrc/group/acme/ac.onguba/acme_scratch/anvil/HOMME_P24.f19_g16_rx1.A.anvil_intel.C.20201008_234625_gl6kf4/bld/tests/thetah-sl-test11conv-r1t2-cdr20/thetah-sl-test11conv-r1t2-cdr20-run.sh
test thetah-sl-test11conv-r1t2-cdr20-run was run successfully
...
file = dcmip2012_test1_1_conv1.nc
The files are different or missing: DIFF_RESULT=DIFFERENT
############################################################################
CPRNC returned the following RMS differences
 RMS Q                                3.1290E-07            NORMALIZED  9.4755E-07
 RMS Q2                               9.4254E-09            NORMALIZED  1.1598E-08
 RMS Q3                               9.6229E-16            NORMALIZED  9.3539E-15
 RMS Q4                               3.0139E-09            NORMALIZED  4.8125E-09
 RMS Q                                4.7749E-07            NORMALIZED  1.4439E-06
 RMS Q2                               7.3927E-07            NORMALIZED  9.0995E-07
 RMS Q3                               9.4491E-09            NORMALIZED  9.2288E-08
 RMS Q4                               3.8253E-07            NORMALIZED  6.1064E-07
############################################################################

      Start 22: thetah-sl-test11conv-r0t1-cdr30-rrm
22/23 Test #22: thetah-sl-test11conv-r0t1-cdr30-rrm ...***Failed    3.48 sec
Submitting 1 jobs
Running test thetah-sl-test11conv-r0t1-cdr30-rrm-run ... /lcrc/group/acme/ac.onguba/acme_scratch/anvil/HOMME_P24.f19_g16_rx1.A.anvil_intel.C.20201008_234625_gl6kf4/bld/tests/thetah-sl-test11conv-r0t1-cdr30-rrm/thetah-sl-test11conv-r0t1-cdr30-rrm-run.sh
test thetah-sl-test11conv-r0t1-cdr30-rrm-run was run successfully
...
file = dcmip2012_test1_1_conv1.nc
The files are different or missing: DIFF_RESULT=DIFFERENT
############################################################################
CPRNC returned the following RMS differences
 RMS Q                                1.3064E-06            NORMALIZED  3.9567E-06
 RMS Q2                               7.0464E-07            NORMALIZED  8.6703E-07
 RMS Q3                               7.2474E-09            NORMALIZED  7.0505E-08
 RMS Q4                               1.1924E-07            NORMALIZED  1.9039E-07
 RMS Q                                2.6283E-06            NORMALIZED  7.9526E-06
 RMS Q2                               1.3701E-06            NORMALIZED  1.6862E-06
 RMS Q3                               1.2996E-06            NORMALIZED  1.2662E-05
 RMS Q4                               7.8432E-07            NORMALIZED  1.2523E-06
############################################################################

      Start 23: thetah-sl-dcmip16_test1pg2
23/23 Test #23: thetah-sl-dcmip16_test1pg2 ............***Failed    4.04 sec
Submitting 1 jobs
Running test thetah-sl-dcmip16_test1pg2-run ... /lcrc/group/acme/ac.onguba/acme_scratch/anvil/HOMME_P24.f19_g16_rx1.A.anvil_intel.C.20201008_234625_gl6kf4/bld/tests/thetah-sl-dcmip16_test1pg2/thetah-sl-dcmip16_test1pg2-run.sh
test thetah-sl-dcmip16_test1pg2-run was run successfully
...
file = dcmip2016_test1_pg21.nc
The files are different or missing: DIFF_RESULT=DIFFERENT
############################################################################
CPRNC returned the following RMS differences
 RMS ps                               1.2390E-10            NORMALIZED  1.2390E-15
 RMS T                                5.2080E-12            NORMALIZED  2.2676E-14
 RMS Q                                4.8783E-16            NORMALIZED  2.2687E-13
 RMS Q4                               2.1738E-11            NORMALIZED  9.2141E-06
 RMS Q5                               8.8119E-14            NORMALIZED  1.0739E-07
 RMS ps                               2.2781E-10            NORMALIZED  2.2782E-15
 RMS T                                6.6905E-12            NORMALIZED  2.9132E-14
 RMS Q                                9.8414E-16            NORMALIZED  4.5786E-13
 RMS Q4                               4.2890E-11            NORMALIZED  1.7657E-05
 RMS Q5                               1.9666E-12            NORMALIZED  2.5026E-06
############################################################################
```
since xAVX flag was replaced with xCORE-AVX2 . Diffs are approved by AMB.

**Comment by oksanaguba** [comment]
For HOMMEBFB suite there are more diffs because of HOMMEXX_BFB_TESTING code in many files (more on that at confluence). Failed tests are from the run

```
The following tests FAILED:
          1 - verifyBaselineResults (Failed)
         11 - thetah-test22 (Failed)
         12 - thetanh-test22 (Failed)
         14 - thetah-TC-ftype4 (Failed)
         15 - thetah-TC (Failed)
         16 - thetanh-TC (Failed)
         17 - thetanh-TC-nudiv (Failed)
         18 - thetanh-TC-nutop (Failed)
         19 - thetanhwet-TC (Failed)
         23 - thetah-sl-dcmip16_test1pg2 (Failed)
         42 - theta-f0-tt5-hvs1-hvst0-r3-qz1-nutopoff-ne2-nu3.4e18-ndays1 (Failed)
         43 - theta-f0-tt5-hvs1-hvst0-r3-qz1-nutopoff-kokkos-ne2-nu3.4e18-ndays1 (Failed)
         44 - theta-f1-tt5-hvs1-hvst0-r3-qz1-nutopoff-ne2-nu3.4e18-ndays1 (Failed)
         45 - theta-f1-tt5-hvs1-hvst0-r3-qz1-nutopoff-kokkos-ne2-nu3.4e18-ndays1 (Failed)
         46 - theta-f1-tt5-hvs1-hvst0-r0-qz1-nutopoff-ne2-nu3.4e18-ndays1 (Failed)
         47 - theta-f1-tt5-hvs1-hvst0-r0-qz1-nutopoff-kokkos-ne2-nu3.4e18-ndays1 (Failed)
         48 - theta-f1-tt10-hvs1-hvst0-r3-qz1-nutopoff-ne2-nu3.4e18-ndays1 (Failed)
         49 - theta-f1-tt10-hvs1-hvst0-r3-qz1-nutopoff-kokkos-ne2-nu3.4e18-ndays1 (Failed)
         50 - theta-f1-tt10-hvs1-hvst0-r2-qz10-nutopoff-GB-ne2-nu3.4e18-ndays1 (Failed)
         51 - theta-f1-tt10-hvs1-hvst0-r2-qz10-nutopoff-GB-kokkos-ne2-nu3.4e18-ndays1 (Failed)
         57 - theta-f0-tt5-hvs1-hvst0-r3-qz1-nutopoff-ne6-nu1.25e17-ndays1 (Failed)
         58 - theta-f0-tt5-hvs1-hvst0-r3-qz1-nutopoff-kokkos-ne6-nu1.25e17-ndays1 (Failed)
         59 - theta-f1-tt5-hvs1-hvst0-r3-qz1-nutopoff-ne6-nu1.25e17-ndays1 (Failed)
         60 - theta-f1-tt5-hvs1-hvst0-r3-qz1-nutopoff-kokkos-ne6-nu1.25e17-ndays1 (Failed)
         61 - theta-f1-tt5-hvs1-hvst0-r0-qz1-nutopoff-ne6-nu1.25e17-ndays1 (Failed)
         62 - theta-f1-tt5-hvs1-hvst0-r0-qz1-nutopoff-kokkos-ne6-nu1.25e17-ndays1 (Failed)
         63 - theta-f1-tt10-hvs1-hvst0-r3-qz1-nutopoff-ne6-nu1.25e17-ndays1 (Failed)
         64 - theta-f1-tt10-hvs1-hvst0-r3-qz1-nutopoff-kokkos-ne6-nu1.25e17-ndays1 (Failed)
         65 - theta-f1-tt10-hvs1-hvst0-r2-qz10-nutopoff-GB-ne6-nu1.25e17-ndays1 (Failed)
         66 - theta-f1-tt10-hvs1-hvst0-r2-qz10-nutopoff-GB-kokkos-ne6-nu1.25e17-ndays1 (Failed)
```
Many fails are just new tests, all unit tests pass, old F theta tests fail due to BFB_TESTING changes.
There is also a side behavior that F theta execs now have xCORE_AVX2 flag (but nonbfb builds do not), we'll ignore that for now.

Again, for the new code within BFB_TESTING, more will be on confluence.

**Comment by oksanaguba** [comment]
We also ran atm_integration, all pass:

```
[ac.onguba@blueslogin2 as]$ ./cs.status.20201009_003757_ig54k0 | grep Overall
  ERP_Ln9.ne4_ne4.F-EAMv1-AQP1.anvil_intel (Overall: PASS) details:
  ERP_Ln9.ne4_ne4.F-EAMv1-RCEMIP.anvil_intel (Overall: PASS) details:
  ERS_Ld5.ne4_ne4.FC5AV1C-L.anvil_intel.eam-gust_param (Overall: PASS) details:
  ERS_Ld5.ne4_ne4.FC5AV1C-L.anvil_intel.eam-rrtmgp (Overall: PASS) details:
  ERS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetahy_ftype2 (Overall: PASS) details:
  ERS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetanh_ftype2 (Overall: PASS) details:
  PEM_Ln5.ne4_ne4.FC5AV1C-L.anvil_intel (Overall: PASS) details:
  PET_Ln5.ne4_ne4.FC5AV1C-L.anvil_intel.allactive-mach-pet (Overall: PASS) details:
  REP_Ln5.ne4_ne4.FC5AV1C-L.anvil_intel (Overall: PASS) details:
  SMS_D_Ln5.ne4_ne4.FC5AV1C-L.anvil_intel.eam-cosplite_nhtfrq5 (Overall: PASS) details:
  SMS_Ld1.ne4_ne4.F-EAMv1-AQP1.anvil_intel.eam-clubb_only (Overall: PASS) details:
  SMS_Ld9.ne4pg2_ne4pg2.FC5AV1C-04P2.anvil_intel.eam-thetahy_sl_pg2_mass (Overall: PASS) details:
  SMS_Ln1.ne4_ne4.FC5AV1C-L.anvil_intel.eam-chem_pp (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-preqx_ftype0 (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-preqx_ftype1 (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-preqx_ftype4 (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetahy_ftype0 (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetahy_ftype1 (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetahy_ftype2 (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetahy_ftype4 (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetahy_sl (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetanh_ftype0 (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetanh_ftype1 (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetanh_ftype2 (Overall: PASS) details:
  SMS.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetanh_ftype4 (Overall: PASS) details:
[ac.onguba@blueslogin2 as]$ ./cs.status.20201009_003757_ig54k0 | grep FAIL
    FAIL ERS_Ld5.ne4_ne4.FC5AV1C-L.anvil_intel.eam-gust_param TPUTCOMP Error: Computation time increase > 10 pct from baseline
    FAIL ERS_Ld5.ne4_ne4.FC5AV1C-L.anvil_intel.eam-rrtmgp TPUTCOMP Error: Computation time increase > 10 pct from baseline
    FAIL SMS_D_Ln5.ne4_ne4.FC5AV1C-L.anvil_intel.eam-cosplite_nhtfrq5 TPUTCOMP Error: Computation time increase > 10 pct from baseline
    FAIL SMS_Ld1.ne4_ne4.F-EAMv1-AQP1.anvil_intel.eam-clubb_only TPUTCOMP Error: Computation time increase > 10 pct from baseline
```

**Comment by oksanaguba** [comment]
Last one is a _D case:

```
[ac.onguba@blueslogin2 as]$ ./cs.status.20201009_121743_v7f2xl | grep Overall
  SMS_D.ne4_ne4.FC5AV1C-L.anvil_intel.eam-thetanh_ftype4 (Overall: PASS) details:
```

**Comment by oksanaguba** [comment]
Confluence for BFB_TESTING embedded code https://acme-climate.atlassian.net/wiki/spaces/NGDNA/pages/1899823175/Integration+of+NH+thetaxx+to+master+verifying . All looks ok.

**Comment by oksanaguba** [comment]
CDash after merge to next https://my.cdash.org/index.php?project=E3SM&date=2020-10-11 .

**Pull request merged**